src/main/java/com/ldtteam/structurize/placement/StructurePlacer.java

```diff
@@ -289,18 +290,26 @@ public BlockPlacementResult handleBlockPlacement(
             worldEntity = world.getBlockEntity(worldPos);
         }
 
-        if (BlockUtils.areBlockStatesEqual(localState, worldState, handler::replaceWithSolidBlock, handler.fancyPlacement(), handler::shouldBlocksBeConsideredEqual, tileEntityData, worldEntity))
-        {
-            return new BlockPlacementResult(worldPos, BlockPlacementResult.Result.SUCCESS);
-        }
-
         if (localState.getBlock() == ModBlocks.blockSolidSubstitution.get() && handler.fancyPlacement())
         {
             localState = this.handler.getSolidBlockForPos(worldPos, handler.getBluePrint().getRawBlockStateFunction().compose(handler::getStructurePosFromWorld));
         }
         if (localState.getBlock() == ModBlocks.blockTagSubstitution.get() && handler.fancyPlacement())
         {
-            localState = Blocks.AIR.defaultBlockState();
+            if (tileEntityData != null && BlockEntity.loadStatic(localPos, localState, tileEntityData) instanceof BlockEntityTagSubstitution tagEntity)
+            {
+                localState = tagEntity.getReplacement().getBlockState();
+                tileEntityData = tagEntity.getReplacement().getBlockEntityTag();
+            }
+            else
+            {
+                localState = Blocks.AIR.defaultBlockState();
+            }
+        }
+
+        if (BlockUtils.areBlockStatesEqual(localState, worldState, handler::replaceWithSolidBlock, handler.fancyPlacement(), handler::shouldBlocksBeConsideredEqual, tileEntityData, worldEntity))
```

Comment:
why move this down?

Reply:
Because it should calculate the target block of the placeholders before it decides whether or not the target block is already in the world.  Not doing that is why "placeholders" often show up in the resource requirements lists unnecessarily.